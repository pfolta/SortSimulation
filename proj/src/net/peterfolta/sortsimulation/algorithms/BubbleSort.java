/*
 * SortSimulation - A visual representation of sorting algorithms
 * Copyright (C) 2008-2015 Peter Folta. All rights reserved.
 * 
 * Project:			SortSimulation 
 * Version:			2.0.1
 * Website:			http://www.peterfolta.net/software/sortsimulation
 * 
 * File:			Bubblesort.java
 * Created:			2008/11/29
 * Last modified:	2015/2/16
 * Author:			Peter Folta <mail@peterfolta.net>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.peterfolta.sortsimulation.algorithms;

import net.peterfolta.sortsimulation.common.ArrayTools;
import net.peterfolta.sortsimulation.common.interfaces.Sortable;
import net.peterfolta.sortsimulation.main.Main;

public class BubbleSort implements Sortable {
	
	private boolean interrupted = false;
	
	public void sort(int[] a, final int index) {
		for (int i = a.length-1; i >= 0 && !interrupted; i--) {
			for (int j = 0; j <= i-1 && !interrupted; j++) {
				if (a[j] > a[j+1]) {
					ArrayTools.swap(a, j, j+1);
				}
				
				try {
					Thread.sleep(Main.settings.getDelay().getDelay());
				} catch (InterruptedException exception) {
					interrupted = true;
					break;
				}
				
				Main.getGUI().getDisplay().asyncExec(new Runnable() {
					public void run() {
						Main.getGUI().getMainWindow().repaintCanvas(index);
					}
				});
			}
		}
	}

}