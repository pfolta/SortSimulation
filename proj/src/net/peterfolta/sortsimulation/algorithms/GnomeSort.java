/*
 * SortSimulation - A visual representation of sorting algorithms
 * Copyright (C) 2008-2015 Peter Folta. All rights reserved.
 * 
 * Project:			SortSimulation 
 * Version:			2.0.1
 * Website:			http://www.peterfolta.net/software/sortsimulation
 * 
 * File:			Gnomesort.java
 * Created:			2014/3/22
 * Last modified:	2015/2/16
 * Author:			Peter Folta <mail@peterfolta.net>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.peterfolta.sortsimulation.algorithms;

import net.peterfolta.sortsimulation.common.ArrayTools;
import net.peterfolta.sortsimulation.common.interfaces.Sortable;
import net.peterfolta.sortsimulation.main.Main;

public class GnomeSort implements Sortable {
	
	private boolean interrupted = false;
	
	public void sort(int[] a, final int index) {
		int pos = 1;
		
		while (pos < a.length && !interrupted) {
			if (a[pos] >= a[pos-1]) {
				pos++;
			} else {
				ArrayTools.swap(a, pos, pos-1);
				
				if (pos > 1) {
					pos--;
				}
			}
			
			try {
				Thread.sleep(Main.settings.getDelay().getDelay());
			} catch (InterruptedException exception) {
				interrupted = true;
				break;
			}
			
			Main.getGUI().getDisplay().asyncExec(new Runnable() {
				public void run() {
					Main.getGUI().getMainWindow().repaintCanvas(index);
				}
			});
		}
	}

}