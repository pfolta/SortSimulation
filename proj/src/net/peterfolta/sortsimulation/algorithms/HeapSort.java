/*
 * SortSimulation - A visual representation of sorting algorithms
 * Copyright (C) 2008-2015 Peter Folta. All rights reserved.
 * 
 * Project:			SortSimulation 
 * Version:			2.0.1
 * Website:			http://www.peterfolta.net/software/sortsimulation
 * 
 * File:			Heapsort.java
 * Created:			2008/11/29
 * Last modified:	2015/2/16
 * Author:			Peter Folta <mail@peterfolta.net>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.peterfolta.sortsimulation.algorithms;

import net.peterfolta.sortsimulation.common.ArrayTools;
import net.peterfolta.sortsimulation.common.interfaces.Sortable;
import net.peterfolta.sortsimulation.main.Main;

public class HeapSort implements Sortable {
	
	private boolean interrupted = false;
	
	private void sift(int[] a, int l, int r, final int index) {
		int i;
		int j;
		int x;
		
		i = l;
		x = a[l];
		j = 2 * i + 1;
		
		if ((j < r) && (a[j+1] > a[j])) {
			j++;
		}
		
		while ((j <= r) && (a[j] > x) && !interrupted) {
			a[i] = a[j];
			i = j;
			j = j*2+1;
			
			if ((j < r) && (a[j+1] > a[j])) {
				j++;
			}
			
			try {
				Thread.sleep(Main.settings.getDelay().getDelay());
			} catch (InterruptedException exception) {
				interrupted = true;
				break;
			}
			
			Main.getGUI().getDisplay().asyncExec(new Runnable() {
				public void run() {
					Main.getGUI().getMainWindow().repaintCanvas(index);
				}
			});
		}
		
		a[i] = x;
	}
	
	public void sort(int[] a, final int index) {
		int l;
		int r;
		
		for (l = (a.length - 2) / 2; l >= 0 && !interrupted; l--) {
			sift(a, l, a.length-1, index);
		}
		
		for (r = a.length - 1; r > 0 && !interrupted; r--) {
			ArrayTools.swap(a, 0, r);
			sift(a, 0, r-1, index);
		}
	}

}