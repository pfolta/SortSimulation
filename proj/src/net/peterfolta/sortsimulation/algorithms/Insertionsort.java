/*
 * SortSimulation - A visual representation of sorting algorithms
 * Copyright (C) 2008-2014 Peter Folta. All rights reserved.
 * 
 * Project:			SortSimulation 
 * Version:			2.0.0
 * Website:			http://www.peterfolta.net/software/sortsimulation
 * 
 * File:			Insertionsort.java
 * Created:			2008/11/29
 * Last modified:	2014/03/21
 * Author:			Peter Folta <mail@peterfolta.net>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.peterfolta.sortsimulation.algorithms;

import net.peterfolta.sortsimulation.main.Main;

public class Insertionsort {
	
	private boolean interrupted = false;
	
	public void sort(int[] a, final int index) {
		int tmp;
		int j;
		
		for(int i = 1; i < a.length; i++) {
			if(interrupted) {
				break;
			}
			
			tmp = a[i];
			j = i;
			
			while(j > 0 && a[j-1] > tmp) {
				a[j] = a[j-1];
				j--;
				
				try {
					Thread.sleep(Main.settings.getSimulationSpeed());
				} catch (InterruptedException exception) {
					interrupted = true;
					break;
				}
				
				Main.getGUI().getDisplay().asyncExec(new Runnable() {
					public void run() {
						Main.getGUI().getMainWindow().repaintCanvas(index);
					}
				});
			}
			
			a[j] = tmp;
		}
	}

}