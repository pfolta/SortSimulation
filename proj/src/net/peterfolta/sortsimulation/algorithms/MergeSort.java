/*
 * SortSimulation - A visual representation of sorting algorithms
 * Copyright (C) 2008-2015 Peter Folta. All rights reserved.
 * 
 * Project:			SortSimulation 
 * Version:			2.0.1
 * Website:			http://www.peterfolta.net/software/sortsimulation
 * 
 * File:			Mergesort.java
 * Created:			2008/11/30
 * Last modified:	2015/2/16
 * Author:			Peter Folta <mail@peterfolta.net>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.peterfolta.sortsimulation.algorithms;

import net.peterfolta.sortsimulation.common.interfaces.Sortable;
import net.peterfolta.sortsimulation.gui.GUI;
import net.peterfolta.sortsimulation.main.Main;

public class MergeSort implements Sortable {
	
	private int[] a;
	private int[] b;
	private int n;
	
	private boolean interrupted = false;
	
	public void sort(int[] a, final int index) {
		this.a = a;
		n = a.length;
		b = new int[n];
		mergesort(0, n-1, index);
	}
	
	private void mergesort(int lo, int hi, final int index) {
		if (lo < hi) {
			int m = (lo + hi) / 2;
			mergesort(lo, m, index);
			mergesort(m+1, hi, index);
			merge(lo, m, hi, index);
		}
	}
	
	private void merge(int lo, int m, int hi, final int index) {
		int i = lo;
		int j = hi;
		int k = lo;
		
		while (i <= m && !interrupted) {
			b[k++] = a[i++];
		}
		
		while (j > m && !interrupted) {
			b[k++] = a[j--];
		}
		
		i = lo;
		j = hi;
		k = lo;
		
		while (i <= j && !interrupted) {
			if (b[i] <= b[j]) {
				a[k++] = b[i++];
			} else {
				a[k++] = b[j--];
			}
			
			try {
				Thread.sleep(Main.settings.getDelay().getDelay());
			} catch (InterruptedException exception) {
				interrupted = true;
				break;
			}
			
			GUI.getInstance().getDisplay().asyncExec(new Runnable() {
				public void run() {
					GUI.getInstance().getMainWindow().repaintCanvas(index);
				}
			});
		}
	}

}