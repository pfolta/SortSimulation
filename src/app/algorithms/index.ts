export type { default as SortingAlgorithm } from "@/app/algorithms/SortingAlgorithm";
export { default as BubbleSort } from "@/app/algorithms/BubbleSort";
export { default as SelectionSort } from "@/app/algorithms/SelectionSort";
