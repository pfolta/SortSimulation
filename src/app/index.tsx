import React from "react";
import { render } from "react-dom";
import { setAppElement } from "react-modal";

import App from "@/app/App";

const appElement = document.querySelector("#app") as HTMLElement;

setAppElement(appElement);
render(<App />, appElement);
