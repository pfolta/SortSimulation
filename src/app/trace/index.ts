export type { default as Phase } from "@/app/trace/Phase";
export * from "@/app/trace/Phase";

export type { default as Group } from "@/app/trace/Group";
export * from "@/app/trace/Group";

export * from "@/app/trace/Trace";
