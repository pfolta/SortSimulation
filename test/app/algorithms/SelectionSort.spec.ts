import { SelectionSort } from "@/app/algorithms";
import { GroupId } from "@/app/trace";
import "@/app/utils/Array";

describe("Selection Sort", () => {
    describe("sort", () => {
        const array = [6, 5, 10, 9, 8, 4, 7, 3, 1, 2];

        it("returns a trace starting with the initial array", () => {
            const trace = SelectionSort.sort([...array]);
            expect(trace.first().array).toEqual(array);
        });

        it("returns a trace containing the correct array steps", () => {
            const trace = SelectionSort.sort([...array]);
            expect(trace.map((step) => step.array)).toMatchSnapshot();
        });

        it("returns a trace ending with the sorted array", () => {
            const trace = SelectionSort.sort([...array]);
            expect(trace.last().array.isSorted()).toBeTrue();
            expect(trace.last().groups.find((group) => group.id === GroupId.SORTED)?.indices).toIncludeAllMembers(Array.range(array));
        });
    });
});
