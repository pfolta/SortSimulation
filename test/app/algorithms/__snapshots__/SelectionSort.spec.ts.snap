// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selection Sort sort returns a trace containing the correct array steps 1`] = `
Array [
  Array [
    6,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    1,
    2,
  ],
  Array [
    6,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    1,
    2,
  ],
  Array [
    6,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    1,
    2,
  ],
  Array [
    6,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    1,
    2,
  ],
  Array [
    6,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    1,
    2,
  ],
  Array [
    6,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    1,
    2,
  ],
  Array [
    6,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    1,
    2,
  ],
  Array [
    6,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    1,
    2,
  ],
  Array [
    6,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    1,
    2,
  ],
  Array [
    6,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    1,
    2,
  ],
  Array [
    1,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    2,
  ],
  Array [
    1,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    2,
  ],
  Array [
    1,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    2,
  ],
  Array [
    1,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    2,
  ],
  Array [
    1,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    2,
  ],
  Array [
    1,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    2,
  ],
  Array [
    1,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    2,
  ],
  Array [
    1,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    2,
  ],
  Array [
    1,
    5,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    2,
  ],
  Array [
    1,
    2,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    5,
  ],
  Array [
    1,
    2,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    5,
  ],
  Array [
    1,
    2,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    5,
  ],
  Array [
    1,
    2,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    5,
  ],
  Array [
    1,
    2,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    5,
  ],
  Array [
    1,
    2,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    5,
  ],
  Array [
    1,
    2,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    5,
  ],
  Array [
    1,
    2,
    10,
    9,
    8,
    4,
    7,
    3,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    9,
    8,
    4,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    9,
    8,
    4,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    9,
    8,
    4,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    9,
    8,
    4,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    9,
    8,
    4,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    9,
    8,
    4,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    9,
    8,
    4,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    4,
    8,
    9,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    4,
    8,
    9,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    4,
    8,
    9,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    4,
    8,
    9,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    4,
    8,
    9,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    4,
    8,
    9,
    7,
    10,
    6,
    5,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    9,
    7,
    10,
    6,
    8,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    9,
    7,
    10,
    6,
    8,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    9,
    7,
    10,
    6,
    8,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    9,
    7,
    10,
    6,
    8,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    9,
    7,
    10,
    6,
    8,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    10,
    9,
    8,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    10,
    9,
    8,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    10,
    9,
    8,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    10,
    9,
    8,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    10,
    9,
    8,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    10,
    9,
    8,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
  ],
]
`;
